# Probability of Rolling each 2-dice point among 4 dice

- What is the probability of rolling any particular point value (from 2 to 12) for
  any roll of 4 dice?
  
  
Perhaps the most straightforward way to answer these questions is by exhaustive simulation
of 4 dice rolls (which only require $6^4 = 1296$ rolls).

In [44]:
from itertools import product

In [45]:
dice = (1,2,3,4,5,6)
all_rolls = list(product(dice, dice, dice, dice))
len(all_rolls)

1296

In [46]:
def points(dice):
    """ Return set of point values that can be derived from a given (a pair of dice)
        in a roll of the dice.
    """
    result = set()
    num_dice = len(dice)
    for first in range(num_dice):
        for second in range(first+1, num_dice):
            result.add(dice[first] + dice[second])
            
    return result

In [47]:
all_points = [points(x) for x in all_rolls]

In [48]:
def prob_point(p):
    c = 0
    for points in all_points:
        if p in points:
            c += 1
    return c / len(all_points)

In [49]:
for p in range(2, 13):
    print(f"{p:2d}: {prob_point(p):.1%}")

 2: 13.2%
 3: 23.3%
 4: 35.6%
 5: 44.8%
 6: 56.1%
 7: 64.4%
 8: 56.1%
 9: 44.8%
10: 35.6%
11: 23.3%
12: 13.2%


# Markov Modeling

Alternately, we can calculate the probabilities more directly by modeling
dice rolls with Markov processes and assigning a state to the specific point
value to be rolled.

For example, to model rolling a $2$ from a pair of dice among a sequence of
single-dice rolls a simple model need only consider when a $1$ is rolled, vs.
all the other faces on a die.  There is one way to roll a $1$ and five ways
of rolling another number.

# Roll 2

Looking for (1, 1).

<img src="./images/roll-2.png" width="200">

The state transitions can be modeled by a 3x3 array where the state-to-state
probabilities for the transition from state $i$ to $j$ is $A_{ij}$.

In [50]:
from pandas import DataFrame as df

roll_2 = df(((5/6, 0, 0), (1/6, 5/6, 0), (0, 1/6, 1)))
roll_2

0         1  2
0  0.833333  0.000000  0
1  0.166667  0.833333  0
2  0.000000  0.166667  1

In [51]:
init = df(((1,), (0,), (0,)))
init

0
0  1
1  0
2  0

Multiplying a vector indicating the initial state, will result in the
probabilities of each state after a single roll.

In [52]:
roll_2.dot(init)

0
0  0.833333
1  0.166667
2  0.000000

In [53]:
def matrix_power(m, n):
    result = m
    for _ in range(n-1):
        result = result.dot(m)
    return result

In [54]:
matrix_power(roll_2, 4)

0         1    2
0  0.482253  0.000000  0.0
1  0.385802  0.482253  0.0
2  0.131944  0.517747  1.0

The probability of being in each state after 4 rolls of a die.

In [55]:
matrix_power(roll_2, 4).dot(init)

0
0  0.482253
1  0.385802
2  0.131944

So, the probability of rolling a $2$ with four dices is $0.131944$.

We can come up with similar models for each of the other points (3, 4, 5, 6,
and 7) and note that by symmetry the points, $N$, above 7 are identical to the
models for $7 - N$.

# Roll 3

Looking for (1, 2).

<img src="./images/roll-3.png" width="200">

In [56]:
roll_3 = df(((4/6, 0, 0), (2/6, 5/6, 0), (0, 1/6, 1)))
print(roll_3)
print(matrix_power(roll_3, 4).dot(init))

          0         1  2
0  0.666667  0.000000  0
1  0.333333  0.833333  0
2  0.000000  0.166667  1
          0
0  0.197531
1  0.569444
2  0.233025


# Roll 4

Looking for (1, 3) or (2, 2).

<img src="./images/roll-4.png" width="300">

In [60]:
roll_4 = df((
    (3/6, 0, 0, 0, 0),
    (2/6, 4/6, 0, 0, 0),
    (1/6, 0, 3/6, 0, 0),
    (0, 1/6, 2/6, 4/6, 0),
    (0, 1/6, 1/6, 2/6, 1)))
print(roll_4)
print(matrix_power(roll_4, 4).dot((1, 0, 0, 0, 0)))

          0         1         2         3  4
0  0.500000  0.000000  0.000000  0.000000  0
1  0.333333  0.666667  0.000000  0.000000  0
2  0.166667  0.000000  0.500000  0.000000  0
3  0.000000  0.166667  0.333333  0.666667  0
4  0.000000  0.166667  0.166667  0.333333  1
0    0.062500
1    0.270062
2    0.083333
3    0.228395
4    0.355710
dtype: float64


# Roll 5

Looking for (1, 4) or (2, 3).